### DISCUSSION TITLE
Headless mode with GPUs

### DISCUSSION THREAD

**Discussion by Suke0811** [cyberbotics/webots#5655] [General]
I have tested Webots headless mode in docker as described [here](https://www.cyberbotics.com/doc/guide/installation-procedure?version=develop#installing-the-docker-image) in Ubuntu 20.04.

In the headless mode, if we make gpus available to the docker container as follows
```bash
docker run --gpus=all -it cyberbotics/webots:latest
xvfb-run webots --stdout --stderr --batch --mode=realtime /path/to/your/world/file
```
This seems to launch Webots without GPUs (hardware acceleration). This is, to my knowledge, because xvfb will use software acceleration.

I believe that GPUs will speed up the simulations for cameras and video recordings, so in that case, I'd like to use GPUs in the headless. 

This is the output with xvfb-run webots....
You can see that Webots warns that the system is below minimum. With GUI + GPU in docker, I don't get warnings or it runs nearly as fast as a native installation.

```bash

xvfb-run webots --stdout --stderr --batch --mode=fast /usr/local/webots/projects/robots/clearpath/moose/worlds/moose_demo.wbt
WARNING: It is not recommended to run Webots as root.
WARNING: System below the minimal requirements.

Webots has detected that your GPU vendor is 'Mesa/X.org'. A recent NVIDIA or AMD graphics adapter is highly recommended to run Webots smoothly. Webots has detected that your computer uses a slow 3D software rendering system. It is strongly recommended to install the latest graphics drivers provided by your GPU manufacturer. Webots will run much faster after the installation of the correct driver.

 - Shadows have been deactivated.
 - Anti-aliasing has been deactivated.
 - Main 3D view global ambient occlusion has been de-activated.
 - Texture quality has been reduced.
 - Texture maximum filtering has been reduced due to GPU limitations.
 - Anisotropic texture filtering is not supported by the GPU.

You can try to re-activate some OpenGL features from the Webots preferences.
WARNING: Cannot initialize the sound engine: Cannot initialize OpenAL default device 'OpenAL Soft'
```

Are there any ways of using hardware acceleration in the headless mode?

**Comment by ygoumaz** [comment]
Please ask this question to https://robotics.stackexchange.com/ with the `webots` tag. We will be happy to answer you there.

**Reply by Suke0811** [reply]
Thank you for guiding me. I copy the question. 
https://robotics.stackexchange.com/questions/24174/headless-mode-with-gpus

**Reply by ygoumaz** [reply]
I replied to you, let me know if the proposed solution works for you. If this is the case, you can upvote the answer and accept it so that it becomes more visible to other users!
